### PR TITLE
fix(codepipeline): unhelpful artifact validation messages

### PR DESCRIPTION
**Pull request by rix0rrr**
The artifact validation error messages are pretty unhelpful, just
saying things like "artifact X gets consumed before it gets produced"
(or similar), without actually referencing the stages/actions involved.

This becomes problematic if the pipeline got generated for you by
automation and indirection, because you can't simply grep your codebase
for the offending artifact name.

Make the messages more explicit and clear so it's a lot more obvious
what's going on (and hopefully getting a fighting chance to figure out
what's wrong).

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts [hunk 1]
```diff
@@ -749,34 +749,73 @@ export class Pipeline extends PipelineBase {
...
-
-      for (const action of sortedActions) {
-        // start with inputs
-        const inputArtifacts = action.inputs;
-        for (const inputArtifact of inputArtifacts) {
-          if (!inputArtifact.artifactName) {
-            ret.push(`Action '${action.actionName}' has an unnamed input Artifact that's not used as an output`);
-          } else if (!outputArtifactNames.has(inputArtifact.artifactName)) {
-            ret.push(`Artifact '${inputArtifact.artifactName}' was used as input before being used as output`);
+    interface PipelineLocation {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts:752]: Can we make this a class instead, with behaviors?

```ts
class PipelineLocation {
  public readonly stageName: string;

  constructor(stage: IStage, public readonly action: FullActionDescriptor, public readonly stageIndex: number) {
    this.stageName = stage.stageName;
  }

  public lessThanOrEqual(that: PipelineLocation): boolean {
    return this.stageIndex === that.stageIndex
      ? this.action.runOrder <= that.action.runOrder
      : this.stageIndex < that.stageIndex;
  }

  public toString(): string {
    return `stage ${loc.stageIndex + 1} ('${loc.stageName}') action ${loc.action.runOrder} ('${loc.action.actionName}')`;
  }
}
```

File: packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts [hunk 2]
```diff
@@ -749,34 +749,73 @@ export class Pipeline extends PipelineBase {
...
+      stageIndex: number;
+      stageName: string;
+      action: FullActionDescriptor;
+    }
+
+    /**
+     * Returns whether a is before or the same order as b
+     */
+    function beforeOrEqual(a: PipelineLocation, b: PipelineLocation) {
+      if (a.stageIndex !== b.stageIndex) { return a.stageIndex < b.stageIndex; }
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts:762]: I think it's much simpler to do `if (positiveCondition) sth else sth_else`, then to start with the negative condition. I've proposed a version above.

File: packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts [hunk 3]
```diff
@@ -749,34 +749,73 @@ export class Pipeline extends PipelineBase {
...
+      return a.action.runOrder <= b.action.runOrder;
+    }
+
+    function renderLoc(loc: PipelineLocation) {
+      // runOrders are 1-based, so make the stageIndex also 1-based otherwise it's going to be confusing.
+      return `stage ${loc.stageIndex + 1} ('${loc.stageName}') action ${loc.action.runOrder} ('${loc.action.actionName}')`;
+    }
+
+    const producer: Record<string, PipelineLocation> = {};
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts:771]: Is this really a canonical usage for `Record`...? I feel like `{ [name: string]: PipelineLocation }` is much clearer...

File: packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts [hunk 4]
```diff
@@ -749,34 +749,73 @@ export class Pipeline extends PipelineBase {
...
+    const firstConsumer: Record<string, PipelineLocation> = {};
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts:772]: Can you change these to plurals? So, `producers` and `firstConsumers`. It's actually pretty confusing to have `producer` and `firstConsumer` be collections, but `thisLocation` and `producerLoc` below as single values.

File: packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts [hunk 5]
```diff
@@ -749,34 +749,73 @@ export class Pipeline extends PipelineBase {
...
+    for (const [stageIndex, stage] of enumerate(this._stages)) {
+      // For every output artifact, get the producer
+      for (const action of stage.actionDescriptors) {
+        const thisLocation = { stageIndex, action, stageName: stage.stageName };
+
+        for (const outputArtifact of action.outputs) {
+          // output Artifacts always have a name set
+          const name = outputArtifact.artifactName!;
+          if (producer[name]) {
+            ret.push(`Artifact '${name}' is used as an output by both '${producer[name].action.actionName}' and '${action.actionName}'. Every artifact can only be produced once.`);
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts:783]: Do you want to add the Stage here as well? (Actions don't have to have unique names across different Stages)

File: packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts [hunk 6]
```diff
@@ -749,34 +749,73 @@ export class Pipeline extends PipelineBase {
...
+            continue;
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts:784]: How about:

```ts
if (producer[name]) {
  ret.push(`...`);
} else {
  producer[name] = thisLocation;
}
```

No `continue` necessary.

File: packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts [hunk 7]
```diff
@@ -749,34 +749,73 @@ export class Pipeline extends PipelineBase {
...
-          // output Artifacts always have a name set
-          if (outputArtifactNames.has(outputArtifact.artifactName!)) {
-            ret.push(`Artifact '${outputArtifact.artifactName}' has been used as an output more than once`);
-          } else {
-            outputArtifactNames.add(outputArtifact.artifactName!);
+        // For every input artifact, get the first consumer
+        for (const inputArtifact of action.inputs) {
+          const name = inputArtifact.artifactName;
+          if (!name) {
+            ret.push(`Action '${action.actionName}' has an unnamed input Artifact (probably not used as an output in this pipeline)`);
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts:794]: Why change this message here - why the "probably"? There's nothing probable about that, it's a certainty 🙂. I would say leave it what it was before: `Action '${action.actionName}' has an unnamed input Artifact that's not used as an output`.

File: packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts [hunk 8]
```diff
@@ -749,34 +749,73 @@ export class Pipeline extends PipelineBase {
...
         }
       }
     }
 
+    // Now validate that every input artifact is produced before it's
+    // being consumed.
+    for (const [artifactName, consumerLoc] of Object.entries(firstConsumer)) {
+      const producerLoc = producer[artifactName];
+      if (!producerLoc) {
+        ret.push(`Artifact '${artifactName}' is used as an input by '${consumerLoc.action.actionName}', but is not being produced in this pipeline`);
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts:810]: This phrasing is a little weird... perhaps something like `Artifact '${artifactName}' is used as an input by '${consumerLoc.action.actionName}', but is never used as the output of any action in the pipeline`?

File: packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts (hunk 3 above)
Comment (rix0rrr) [reply packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts:771]: Yeh I personally hate that notation.

File: packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts (hunk 8 above)
Comment (rix0rrr) [reply packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts:810]: So one of these (forgot which) was the one that threw me off.

Both this and the previous error message you commented on actually have the same root cause (input is not being produced), but the error messages are confusingly different. The only difference is one has a name and the other one doesn't, but the user error is exactly the same.

Also "being used as an output" kind implies some optionality to me,  even though there isn't really, and sounds very mechanical. I'd much prefer "is not being produced": it's very clear that you can't consume something that's not being produced.

File: packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts (hunk 2 above)
Comment (rix0rrr) [reply packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts:762]: True, but in the case of a comparator this is much more easily extended by adding in more lines at the appropriate location (either before or after the first condition). Whereas your style will become more complicated.

Now I know that we only have 2 sorting keys, and probably won't ever have more, but I still prefer writing comparators in that style.

Plus, I'm all about the early returns baby!

File: packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts (hunk 4 above)
Comment (rix0rrr) [reply packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts:772]: But... but... but... what about the beautiful duality between a function and a map?

If there had been a function which would give you the producer location from an artifact ID, would you have given it the plural name?

```ts
// This is a function that gets me the value I want
const loc = producer('abc')

// This is a map that gets me the value I want
const loc = producer['abc']

// -----------------------------
// Sooo... you would prefer this?
const loc = producers['abc']

// Would you also write this?
const loc = producers('abc');

// ...and if so, why not? 😛 
```

The only difference is that the value from the map is precomputed, and the value from the function is... well... maybe it's precomputed, maybe it's not, no way to tell. Why would you even care?

Why the preferred focus of the map being a literal bag of values, as opposed to a mapping that can get you the value that you want?

**Comment by rix0rrr** [comment]
Made the error messages action-centric, which I think is more in line with what users generally care about.

File: packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts (hunk 3 above)
Comment (nija-at) [reply packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts]: Ooh an echo (one I've had with Rico elsewhere) 😉

**Pull request merged**